octobot_trading/personal_data/transactions/transactions_manager.py

```diff
@@ -0,0 +1,90 @@
+#  Drakkar-Software OctoBot-Trading
+#  Copyright (c) Drakkar-Software, All rights reserved.
+#
+#  This library is free software; you can redistribute it and/or
+#  modify it under the terms of the GNU Lesser General Public
+#  License as published by the Free Software Foundation; either
+#  version 3.0 of the License, or (at your option) any later version.
+#
+#  This library is distributed in the hope that it will be useful,
+#  but WITHOUT ANY WARRANTY; without even the implied warranty of
+#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+#  Lesser General Public License for more details.
+#
+#  You should have received a copy of the GNU Lesser General Public
+#  License along with this library.
+import collections
+
+import octobot_commons.logging as logging
+
+import octobot_trading.errors as errors
+import octobot_trading.util as util
+
+
+class TransactionsManager(util.Initializable):
+    MAX_TRANSACTIONS_COUNT = 100000
+
+    def __init__(self):
+        super().__init__()
+        self.logger = logging.get_logger(self.__class__.__name__)
+        self.transactions = collections.OrderedDict()
+
+    async def initialize_impl(self):
+        self._reset_transactions()
+
+    def upsert_transaction_instance(self, transaction, replace_if_exists=False):
+        """
+        Add the transaction instance to self.transactions by its transaction_id
+        Can raise DuplicateTransactionIdError if a transaction already exists
+        with the same transaction_id (when not replacing it with :replace_if_exists:)
+        :param transaction:
+        :param replace_if_exists: When True, replaces the transaction if a transaction has the same transaction_id
+        """
+        if transaction.transaction_id not in self.transactions or replace_if_exists:
+            self.transactions[transaction.transaction_id] = transaction
+            self._check_transactions_size()
+        else:
+            raise errors.DuplicateTransactionIdError(
```

Comment:
I think it's just a naming thing: I find it wierd to have an upsert method that raises by default if it's actually doing an update. If we really want these behavior, I think we should name this function `insert_transaction_instance`